### PR TITLE
Dashboard: Replace `query-string` package with `addQueryArgs`

### LINKED ISSUE #8614: Replace usage of `query-string` package with `addQueryArgs`
## Feature Description

Replace usage of `query-string` package with `addQueryArgs` in the design-system package. 

## Additional Context

---

_Do not alter or remove anything below. The following sections will be managed by moderators only._

### PR DESCRIPTION
**Pull request by spacedmonkey**
## Summary

Replace usage of query-string package with addQueryArgs and URLSearchParams.

### QA

- [x] This is a non-user-facing change and requires no QA

This PR can be tested by following these steps:

1.

### UAT

- [x] UAT should use the same steps as above.

This PR can be tested by following these steps:

1.

## Checklist

- [x] This PR addresses an existing issue and I have linked this PR to it in ZenHub
- [x] I have tested this code to the best of my abilities
- [x] I have verified accessibility to the best of my abilities ([docs](https://github.com/google/web-stories-wp/blob/main/docs/accessibility-testing.md))
- [x] I have verified i18n and l10n (translation, right-to-left layout) to the best of my abilities
- [x] This PR contains automated tests (unit, integration, and/or e2e) to verify the code works as intended ([docs](https://github.com/google/web-stories-wp/tree/main/docs#testing))
- [x] I have added documentation where necessary
- [x] I have added a matching `Type: XYZ` label to the PR

---

Fixes #8614

### CODE REVIEW

File: assets/src/dashboard/app/router/routerProvider.js [hunk 1]
```diff
@@ -19,7 +19,6 @@
  */
 import { useRef, useMemo, useEffect, useState } from 'react';
 import PropTypes from 'prop-types';
-import { parse } from 'query-string';
```
Comment (spacedmonkey) [inline assets/src/dashboard/app/router/routerProvider.js]: @BrittanyIRL Can you explain what is happening here? When I called `parse` from query-string I got an empty object every time. Do we know this was working?

**Comment by spacedmonkey** [comment]
The only place we still use query string in PR is here 

https://github.com/google/web-stories-wp/blob/839e85d11d27a2df409817db5b0820cb15546c5d/assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js#L98-L105

Pretty hard to understand what is going on here. Any advice of getting rid of the usage of query string here? CC @barklund @BrittanyIRL

File: package.json [hunk 2]
```diff
@@ -197,6 +196,7 @@
     "stylis-plugin-rtl": "^1.0.0",
     "terser-webpack-plugin": "^4.2.3",
     "thread-loader": "^3.0.4",
+    "query-string": "^7.0.1",
```
Comment (swissspidy) [inline package.json]: I thought the plan was to remove it, not move to dev dependencies?

(Aside: it's not listed in alphabetical order now)

File: assets/src/dashboard/app/router/routerProvider.js (hunk 1 above)
Comment (swissspidy) [reply assets/src/dashboard/app/router/routerProvider.js]: When looking at an individual template on the "Explore Templates" page, the URL will look like `wp-admin/edit.php?post_type=web-story&page=stories-dashboard#/templates-gallery/template-detail?id=3&isLocal=false`

`parse()` turns the `?id=3&isLocal=false` part (`history.current.location.search`) into `{id: "3", isLocal: "false"}`.

So yes, it is working.

File: package.json (hunk 2 above)
Comment (spacedmonkey) [reply package.json]: It is. This is a draft. I need to workout this - https://github.com/google/web-stories-wp/pull/8640#issuecomment-895954084

File: assets/src/dashboard/app/router/routerProvider.js (hunk 1 above)
Comment (spacedmonkey) [reply assets/src/dashboard/app/router/routerProvider.js]: I don't understand what is happening here. Will ask advice from @BrittanyIRL

Comment (BrittanyIRL) [reply assets/src/dashboard/app/router/routerProvider.js]: Pascal's right about what it's for. To expand on it more, take the following: 
I am looking at the detail view of a template. The URL ends with: `/templates-gallery/template-detail?id=3&isLocal=false` 

This is grabbed from history location search of the router (looks like this: `?id=3&isLocal=false`) and is parsed with `parse` to become `{id: "3", isLocal: "false"}` which is stored as `queryParams`. 

`queryParams` is part of the Router's context (`state.queryParams`) and is important to setting what the view is in dashboard/app as well as what template is getting looked at in the detail view. 

So, it is doing something. Removing this will break the template view. 😂  

If you want to remove it we will need to parse the search still.

You can see the queryParam update if you log it like so: 

```
  useEffect(() => {
    console.log(queryParams);
  }, [queryParams]);
```

and go to the templates view and click on a template to go to detail view. 

hope this helps, let me know if you have more questions about it.

**Comment by swissspidy** [comment]
> The only place we still use query string in PR is here
> 
> https://github.com/google/web-stories-wp/blob/839e85d11d27a2df409817db5b0820cb15546c5d/assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js#L98-L105
> 
> Pretty hard to understand what is going on here. Any advice of getting rid of the usage of query string here? CC @barklund @BrittanyIRL

I think this is similar to what is done in the router:

Given the current location like `http://localhost/#/templates-gallery/template-detail?id=3&isLocal=false`, take the hash `/templates-gallery/template-detail?id=3&isLocal=false`, take the query params from there (`?id=3&isLocal=false`) and parse the query params into an object (`{id: 3, isLocal: false }`.

This way the test can confirm whether the hash location was updated properly.

**Comment by BrittanyIRL** [comment]
> The only place we still use query string in PR is here
> 
> https://github.com/google/web-stories-wp/blob/839e85d11d27a2df409817db5b0820cb15546c5d/assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js#L98-L105
> 
> Pretty hard to understand what is going on here. Any advice of getting rid of the usage of query string here? CC @barklund @BrittanyIRL

TL;DR here is we're using it to check that the detail template is updating accurately in accordance with what the URL is.

File: assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js [hunk 3]
```diff
@@ -95,7 +95,8 @@ describe('CUJ: Creator can browse templates in grid view: See pre-built template
   }
 
   function getQueryParams() {
-    const params = new URLSearchParams(window.location.search);
+    const arr = window.location.href.split('?');
```
Comment (swissspidy) [inline assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js]: Using `window.location.href.split('?')` seems brittle.

I know this is just in Karma tests so it works in practice, but this would fail on something like `wp-admin/edit.php?post_type=web-story&page=stories-dashboard#/templates-gallery/template-detail?id=3&isLocal=false` -- see the two question marks.

We want the search params within the `location.hash` part, not `window.location.search`. So a more robust way of writing would be something like this:

```js
const params = (new URL(window.location.hash, window.location.origin)).searchParams;
return Object.fromEntries(params);
```

Comment (spacedmonkey) [reply assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js]: > see the two question marks.

The split is used BECAUSE of the two question marks. In my tests, this was the url `http://localhost:9876/?id=87748993#/templates-gallery/template-detail?id=10&isLocal=false`. There I passed window.location.href, URLSearchParams only return the first question mark. By using split and the popping the last element off, I always got `id=10&isLocal=false` passed to `URLSearchParams`.

I will give your code snippet a try and get back. But split and getting last one worked everywhere I could think.

Comment (spacedmonkey) [reply assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js]: Sadly your snippet didn't work.

Screenshot 2021-08-11 at 10 45 55

Comment (swissspidy) [reply assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js]: Hmm  can you try `window.location.hostname` instead of `window.location.origin`?

Comment (spacedmonkey) [reply assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js]: Didn't work either. 

Screenshot 2021-08-11 at 11 08 41

Comment (swissspidy) [reply assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js]: Ah, of course `host` doesn't have a protocol. Since the base URL is really irrelevant, could just as well hardcode this:


```js
const params = (new URL(window.location.hash, 'http://localhost')).searchParams;
return Object.fromEntries(params);
```

I understand that the `pop()` method works and is clever at that, but it's not immediately obvious from looking at the code what it does. So it's a little hard to grasp.

I know I'm getting nit-picky here but I think using `window.location.hash` like above would make the intention much more clear.

**Review by swissspidy — APPROVED** [review]
LGTM with one nit, as per discussion in the thread

File: assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js (hunk 3 above)
Comment (spacedmonkey) [reply assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js]: > I know I'm getting nit-picky here but I think using window.location.hash like above would make the intention much more clear.

No I agree, I was going to put a comment to explain it. The code before was also super hard for me to understand. 

> const params = (new URL(window.location.hash, 'http://localhost')).searchParams;

I tried this and it didn't work either. I think the issue, is that query string is part of the hash value. These url functions just think it is part of the hash. You have to extract the query from the hash value to work with it. If you want to do some testing, go to `http://localhost:8899/?id=87748993#/templates-gallery/template-detail?id=10&isLocal=false` locally. The page is nothing, but it is the right url structure and can play around with this in the console.

Comment (swissspidy) [reply assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js]: The issue is that the `#` is part of `window.location.hash`.

So this works:

```js
const params = (new URL('/templates-gallery/template-detail?id=10&isLocal=false', 'http://localhost')).searchParams;
console.log(params.get('id')) // 10
```

But this doesn't:

```js
const params = (new URL('#/templates-gallery/template-detail?id=10&isLocal=false', 'http://localhost')).searchParams;
console.log(params.get('id')) // null
```

To remove the hash, could do `window.location.hash.substr(1)`:

```js
const params = (new URL(window.location.hash.substr(1), 'http://localhost')).searchParams;
return Object.fromEntries(params);
```

Comment (spacedmonkey) [reply assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js]: >is that query string is part of the hash value 

Yeah, that is what I thought. 

I am going to put this out there, this isn't super readable to me. 

> const params = (new URL(window.location.hash.substr(1), 'http://localhost')).searchParams;

If the point is to improve readability, then I think what is currently there is better.

Comment (swissspidy) [reply assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js]: Just coming back to the original snippet:

```js
// http://localhost:8899/?id=87748993#/templates-gallery/template-detail?id=10&isLocal=false -> { id: 10, isLocal: false }
const params = (new URL(window.location.hash.substr(1), window.location.origin)).searchParams;
return Object.fromEntries(params);
```

That doesn't look too bad IMO

Comment (spacedmonkey) [reply assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js]: > const params = (new URL(window.location.hash.substr(1), 'http://localhost')).searchParams;

This also didn't work. 
Screenshot 2021-08-11 at 13 40 55

**Review by BrittanyIRL — APPROVED** [review]
lgtm!

File: assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js (hunk 3 above)
Comment (swissspidy) [reply assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js]: Yeah but have you tried my latest suggestion?

Comment (spacedmonkey) [reply assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js]: > Yeah but have you tried my latest suggestion?

Yes :(

Screenshot 2021-08-11 at 23 22 02

Comment (spacedmonkey) [reply assets/src/dashboard/app/views/templateDetails/karma/templateDetails.karma.js]: Works!
